sklearn/cluster/tests/test_optics.py

```diff
@@ -0,0 +1,342 @@
+
+
+def test_optics2():
+    # Tests the optics clustering method and all functions inside it
+    # 'dbscan' mode
+
+    # Compute OPTICS
+    X = [[1, 1]]
+    clust = OPTICS(max_bound=5.0 * 0.3, min_samples=10)
+
+    # Run the fit
+    assert_raises(ValueError, clust.fit, X)
+
+
+def test_empty_extract():
+    # Test extract where fit() has not yet been run.
+    clust = OPTICS(max_bound=5.0 * 0.3, min_samples=10)
+    assert_raises(ValueError, clust.extract_dbscan, 0.01)
+
+
+def test_bad_extract():
+    # Test an extraction of eps too close to original eps
+    centers = [[1, 1], [-1, -1], [1, -1]]
+    X, labels_true = make_blobs(n_samples=750, centers=centers,
+                                cluster_std=0.4, random_state=0)
+
+    # Compute OPTICS
+    clust = OPTICS(max_bound=5.0 * 0.003, min_samples=10)
+    clust2 = clust.fit(X)
+    assert_raises(ValueError, clust2.extract_dbscan, 0.3)
+
+
+def test_close_extract():
+    # Test extract where extraction eps is close to scaled epsPrime
+
+    centers = [[1, 1], [-1, -1], [1, -1]]
+    X, labels_true = make_blobs(n_samples=750, centers=centers,
+                                cluster_std=0.4, random_state=0)
+
+    # Compute OPTICS
+    clust = OPTICS(max_bound=1.0, min_samples=10)
+    clust3 = clust.fit(X)
+    # check warning when centers are passed
+    assert_warns(RuntimeWarning, clust3.extract_dbscan, .3)
+    assert_equal(max(clust3.extract_dbscan(.3)[1]), 3)
+
+
+def test_auto_extract_hier():
+    # Generate sample data
+
+    np.random.seed(0)
+    n_points_per_cluster = 250
+
+    X = np.empty((0, 2))
+    X = np.r_[X, [-5, -2] + .8 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [4, -1] + .1 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [1, -2] + .2 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [-2, 3] + .3 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [3, -2] + 1.6 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [5, 6] + 2 * np.random.randn(n_points_per_cluster, 2)]
+
+    # Compute OPTICS
+
+    clust = OPTICS(min_samples=9)
+
+    # Run the fit
+    clust.fit(X)
+
+    assert_equal(len(set(clust.labels_)), 6)
+
+
+def test_reach_dists():
+    # Tests against known extraction array
+
+    np.random.seed(0)
+    n_points_per_cluster = 250
+
+    X = np.empty((0, 2))
+    X = np.r_[X, [-5, -2] + .8 * np.random.randn(n_points_per_cluster, 2)]
+    # eps not used for 'auto' extract
+    X = np.r_[X, [4, -1] + .1 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [1, -2] + .2 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [-2, 3] + .3 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [3, -2] + 1.6 * np.random.randn(n_points_per_cluster, 2)]
+    X = np.r_[X, [5, 6] + 2 * np.random.randn(n_points_per_cluster, 2)]
+
+    # Compute OPTICS
+
+    clust = OPTICS(min_samples=10, metric='minkowski')
+
+    # Run the fit
+    clust.fit(X)
+
+    # Expected values
```

Comment:
from where?

Reply:
This was added since just checking the # of clusters is too corse; you can have the reachability distances change, but still extract the same number of clusters... testing the distances separately also isolates what specifically breaks (i.e., did the reachability graph extraction break, or did the extraction function.)

I added this a while ago... but the array was calculated from this implementation of OPTICS circa 2015, which was then validated against Brian Clowers implementation (http://chemometria.us.edu.pl/download/optics.py). Running Brian's code on the same input array gives, within machine precision, the same results. Added comment in code.